### PR TITLE
refactor(components): [switch] switch to script-setup syntax

### LINKED ISSUE #6114: [Refactor] Refactor Plan
## Plans

- remove all `@ts-nocheck`
- [`script-setup`](https://github.com/element-plus/element-plus/issues/5709)
   - refactor `state` to `ref`
   - expose all things user need
- [style namespace](https://github.com/element-plus/element-plus/issues/5374)
- [export component instance type](https://github.com/element-plus/element-plus/issues/4308)
- [refactor plan](https://github.com/element-plus/element-plus/pull/3113)
- [JSX](https://github.com/element-plus/element-plus/issues/5708)
   - Unit Test included
- Fix ALL TS errors, and add component path to `build/type-safe.json` and `build/type-safe-stricter.json`.

Please community contributors at me in your PR.

## Progress

- [x] affix
  - #6065
  - #6159
  - #6295
- [x] alert
  - #6066
  - #7973 
- [x] autocomplete
  - #6067
- [x] avatar
  - #6070
- [x] backtop
  - #6069
  - #6160
- [x] badge
  - #6068
  - #6247
- [x] breadcrumb / breadcrumb-item
  - #6073
- [x] button
  - #5933
- [x] button-group
  - #6256
- [x] calendar
  - #6682
- [x] card
  - #6071
- [x] carousel / carousel-item
  - #6681
- [ ] cascader / cascader-panel
- [x] check-tag
  - #6296
- [ ] checkbox / checkbox-button / checkbox-group
  - Assignee: @chenxch 
  - #7794 
- [x] col
  - #6303
  - assignee @HeftyKoo
- [x] collapse / collapse-item
  - #6307
- [x] collapse-transition
  - Assignee: @ryuhangyeong
  - #7953
- [x] collection
  - Assignee: @metanas
  - #7710
- [x] color-picker
  - #9672 
- [x] config-provider
  - Assignee: @holazz 
- [x] container / aside / footer / header / main
  - Assignee: @ryuhangyeong
  - #7725
  - #7707
- [ ] date-picker
  - Assignee: @JeremyWuuuuu 
- [ ] descriptions / descriptions-item
  - #8348
  - #8349
- [x] dialog
  - #6087
  - #6211
  - #6300
- [x] divider
  - #6556
- [ ] drawer [WIP]
  - assignee: @JeremyWuuuuu 
  - #6239
- [ ] dropdown / dropdown-item / dropdown-menu
- [x] empty
  - #6703
- [ ] focus-trap
- [x] form / form-item
  - assignees: @sxzz
  - #5401
- [x] icon
  - #5934
- [x] image
  - #6704 
- [x] image-viewer
  - #6704 
- [ ] infinite-scroll
- [x] input
  - #6545 
- [x] input-number
  - Assignee: @holazz
  - #7888
- [x] link
  - Assignee: @sxzz 
  - #6543
- [ ] loading
  - Assignee: @sxzz
- [ ] menu / menu-item / menu-item-group / sub-menu
- [ ] message
  - Assignee: @JeremyWuuuuu 
- [ ] message-box
  - Assignee: @JeremyWuuuuu 
- [ ] notification
  - Assignee: @JeremyWuuuuu 
- [ ] option / option-group
   - Assignee: @josonho
- [ ] overlay
  - Assignee: @ryuhangyeong
- [x] page-header
  - #6321
- [ ] pagination
  - Assignee: @metanas
  - #7730 
- [x] popconfirm
  - #7522
- [ ] popover
- [x] popper
  - #6725 
- [ ] progress
  - #7795
- [x] radio / radio-button / radio-group
  - #7701 
- [x] rate
  - #6565
- [x] result
  - Assignee: @ryuhangyeong
  - #7891
- [x] row
  - Assignee @HeftyKoo
  - #7961 
- [ ] roving-focus-group
- [x] row
  - #6354
- [x] scrollbar
  - Assignee: @holazz 
  - #7986
- [ ] select
- [ ] select-v2
- [x] skeleton / skeleton-item
  - #6395
- [x] slider
  - Assignee: @holazz 
  - #8008
- [x] slot
  - #6302
- [x] space
  - #8386 
- [x] step / steps
  - Assignee @ryuhangyeong
  - #7783
- [ ] switch
  - Assignee: @chenxch 
  - #7792  
- [ ] tabs / tab-pane
  - #6305
- [ ] ~~table / table-column~~
- [x] tag
  - #6288
- [x] teleport
  - #6301
- [ ] time-picker
- [ ] time-select
  - #7833
- [x] timeline / timeline-item
  - Assignee: @metanas
  - #7705 
- [x] tooltip
  - #9685 
- [ ] transfer
- [ ] tree
- [ ] tree-v2
- [x] upload
  - #6014
- [ ] virtual-list
- [ ] visual-hidden

### PR DESCRIPTION
**Pull request by chenxch**
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by JeremyWuuuuu** [comment]
@chenxch Don't do 

```
- fix xxx
```

in your description because that will close the issue 😢 

You just mention it in the comment would be fin.e

File: packages/components/switch/src/switch.vue [hunk 1]
```diff
@@ -100,162 +100,145 @@ import {
...
+
+onMounted(() => {
+  if (props.activeColor || props.inactiveColor || props.borderColor) {
+    setBackgroundColor()
+  }
+
+  input.value!.checked = checked.value
 })
+
+defineExpose({ focus })
```
Comment (JeremyWuuuuu) [inline packages/components/switch/src/switch.vue:243]:
```suggestion

defineExpose({
  /**
   *  @description manual focus to the switch component
   **/
  focus
})
```

**Comment by chenxch** [comment]
> @chenxch Don't do
> 
> ```
> - fix xxx
> ```
> 
> in your description because that will close the issue 😢
> 
> You just mention it in the comment would be fin.e

thx, I will pay attention to the description later.

**Review by jw-foss — COMMENTED** [review]
I left a few comments for you to consider. 🏓  back to you

File: packages/components/switch/src/switch.vue [hunk 2]
```diff
@@ -100,162 +100,145 @@ import {
...
-      ns.is('checked', checked.value),
-    ])
+const props = defineProps(switchProps)
+const emit = defineEmits(switchEmits)
 
-    const coreStyle = computed<CSSProperties>(() => ({
-      width: addUnit(props.width),
-    }))
+const { formItem } = useFormItem()
+const switchDisabled = useDisabled(computed(() => props.loading))
```
[the remark this answers is not in the document]
Comment (chenxch) [reply packages/components/switch/src/switch.vue:115]: done.

File: packages/components/switch/src/switch.vue [hunk 3]
```diff
@@ -100,162 +100,150 @@ import {
...
+const { inputId } = useFormItemInputId(props, {
+  formItemContext: formItem,
+})
 
-    const actualValue = computed(() => {
-      return isModelValue.value ? props.modelValue : props.value
-    })
+const switchSize = useSize()
+const switchDisabled = useDisabled(computed(() => props.loading))
+const isModelValue = ref(props.modelValue !== false)
```
Comment (JeremyWuuuuu) [inline packages/components/switch/src/switch.vue:123]: **Suggestion:** maybe we should rename this to `isControlled`, and the implementation is:

```suggestion
const isModelValue = ref(isBoolean(props.modelValue))
```

Comment (JeremyWuuuuu) [inline packages/components/switch/src/switch.vue:123]: **Suggestion:** maybe we should rename this to `isControlled`, and the implementation is:

```suggestion
const isModelValue = ref(isBoolean(props.modelValue))
```

File: packages/components/switch/src/switch.vue [hunk 4]
```diff
@@ -100,162 +100,150 @@ import {
...
+
+  const { beforeChange } = props
+  if (!beforeChange) {
+    handleChange()
+    return
+  }
+
+  const shouldChange = beforeChange()
+
+  const isExpectType = [isPromise(shouldChange), isBoolean(shouldChange)].some(
```
Comment (JeremyWuuuuu) [inline packages/components/switch/src/switch.vue:197]: **Suggestion:** consider rename this variable `isExpectedType`

**Review by jw-foss — COMMENTED** [review]
The code looks pretty much good to me, left a few comments for you to consider, back to you 🏓

**Comment by chenxch** [comment]
@JeremyWuuuuu I adjusted accordingly and rechecked the logic of the code. In the review of the previous `watch(()=>props.modelValue,...)` section, I thought that the original logic should be maintained, and the change of whether the controllable switch is there, I reopened the previous dialogue.

**Comment by JeremyWuuuuu** [comment]
> @JeremyWuuuuu I adjusted accordingly and rechecked the logic of the code. In the review of the previous `watch(()=>props.modelValue,...)` section, I thought that the original logic should be maintained, and the change of whether the controllable switch is there, I reopened the previous dialogue.

Got it.

File: packages/components/switch/src/switch.vue [hunk 5]
```diff
@@ -120,7 +120,7 @@ const { inputId } = useFormItemInputId(props, {
 
 const switchSize = useSize()
 const switchDisabled = useDisabled(computed(() => props.loading))
-const isModelValue = ref(props.modelValue !== false)
+const isControlled = ref(props.value === false)
```
Comment (JeremyWuuuuu) [inline packages/components/switch/src/switch.vue:123]: **Question:** why do we change here? IMHO, we should not change this at all as you said, we need to restore this

File: packages/components/switch/src/switch.vue [hunk 6]
```diff
@@ -194,10 +194,8 @@ const switchValue = (): void => {
 
   const shouldChange = beforeChange()
 
-  const isExpectType = [isPromise(shouldChange), isBoolean(shouldChange)].some(
-    (i) => i
-  )
-  if (!isExpectType) {
+  const isPromiseOrBool = isPromise(shouldChange) || isBoolean(shouldChange)
```
Comment (JeremyWuuuuu) [inline packages/components/switch/src/switch.vue:197]: const isPromiseOrBool = [isPromise(shouldChange), isBoolean(shouldChange].includes(true)

**Comment by chenxch** [comment]
cc @jw-foss

**Review by sxzz — COMMENTED** [review]
Hi @chenxch Could you please rebase your code and I will review this PR.

**Review by sxzz — APPROVED** [review]

**Review by sxzz — APPROVED** [review]

**Pull request merged**